pkg/stats/io_snapshots.go

```diff
@@ -37,25 +42,78 @@ func (h *snapshotHistory) SetInput(id string, latestInput json.Marshaler) {
 	h.latestInput[id] = latestInput
 }
 
-func (h *snapshotHistory) GetInput() ([]byte, error) {
+func (h *snapshotHistory) GetInput(format string, clusters []string, namespaces []string, resourceTypes []string) ([]byte, error) {
 	h.lock.RLock()
 	defer h.lock.RUnlock()
 	if h.latestInput == nil {
 		return []byte{}, nil
 	}
-	result := map[string]interface{}{}
-	for id, obj := range h.latestInput {
-		jsn, err := obj.MarshalJSON()
-		if err != nil {
-			return nil, err
+	genericMaps, err := getGenericMaps(h.latestInput)
+	if err != nil {
+		return nil, err
+	}
+
+	translator := genericMaps["translator"].(map[string]interface{})
+	// Filter resource types
+	if resourceTypes != nil {
+		for k := range translator {
+			if !contains(resourceTypes, k) {
+				delete(translator, k)
+			}
 		}
-		genericMap := map[string]interface{}{}
-		if err := json.Unmarshal(jsn, &genericMap); err != nil {
-			return nil, err
+	}
+
+	// Filter cluster and namespaces
+	if clusters != nil || namespaces != nil {
+		for k, resources := range translator {
+			newResources := []interface{}{}
+			for _, resource := range resources.([]interface{}) {
+				resource := resource.(map[string]interface{})
+				metadata := resource["metadata"].(map[string]interface{})
+				cluster, ok := metadata["clusterName"].(string)
+				include := true
+				if clusters != nil && ok && !contains(clusters, cluster) {
+					include = false
+				}
+				namespace, ok := metadata["namespace"].(string)
+				if k == "/v1, Kind=Namespace" {
+					namespace, ok = metadata["name"].(string)
+				}
+				if namespaces != nil && ok && !contains(namespaces, namespace) {
+					include = false
+				}
+				if include {
+					newResources = append(newResources, resource)
+				}
+			}
+			if len(newResources) == 0 {
+				delete(translator, k)
+			} else {
+				translator[k] = newResources
+			}
 		}
-		result[id] = genericMap
 	}
-	return json.Marshal(result)
+
+	// If filters enabled only return translator section
+	if clusters != nil || namespaces != nil || resourceTypes != nil {
+		return formatMap(format, translator)
+	} else {
```

Comment:
Should this just short-circuit all of the filtering logic and return early if it can?